dags/python_scripts/atd_vzd_cr3_scan_pdf_records.py

```diff
@@ -0,0 +1,300 @@
+        return 0
+
+
+def get_timestamp() -> str:
+    """
+    Returns a string containing the date and time the file was tested
+    :return str:
+    """
+    return datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+
+
+def get_file_metadata(crash_id: int) -> dict:
+    """
+    Returns a dictionary containing the metadata for a file. It assumes the file already exists in disk.
+    :param int crash_id:
+    :return dict:
+    """
+
+    if not is_crash_id(crash_id) or not file_exists(crash_id):
+        return {}
+
+    timestamp = get_timestamp()
+    file_size = get_file_size(crash_id)
+    mime_attr = get_mime_attributes(crash_id)
+
+    return {
+        "last_update": timestamp,
+        "file_size": file_size,
+        "mime_type": mime_attr.get("mime_type", None),
+        "encoding": mime_attr.get("encoding", None)
+    }
+
+
+def process_record(crash_id: int) -> bool:
+    """
+    Controls the process for a single file
+    """
+    if not is_crash_id(crash_id):
+        print(f"Invalid crash_id: {crash_id}")
+        return False
+
+    # 1. Download file to disk
+    if not download_file(crash_id):
+        print(f"Could not find CR3 file for crash_id: {crash_id}")
+        return False
+
+    if not file_exists(crash_id):
+        print(f"The PDF file for crash_id: {crash_id} does not exist. Skipping.")
+        return False
+
+    # 2. Generate file metadata
+    metadata = get_file_metadata(crash_id)
+    if not is_valid_metadata(metadata):
+        print(f"Invalid metadata for file for crash_id: {crash_id}")
+        return False
+
+    # 3. Execute GraphQL with new metadata
+    update_metadata(crash_id, metadata)
+
+    # 4. Delete the file from disk
+    delete_file(crash_id)
+    return True
+
+
+def get_file_metadata_cloud(crash_id: int) -> Optional[dict]:
+    """
+    Attempts to retrieve the metadata from the database, which can be NULL.
+    :param int crash_id: The crash_id to retrieve it from
+    :return dict:
+    """
+    if not is_crash_id(crash_id):
+        # If the crash id is not valid, return None
+        return None
+
+    # Executes GraphQL to fetch any crashes without CR3 PDF metadata, with a limit as indicated
+    query_get_record = """
+        query getUnverifiedCr3($crash_id:Int!) {
+          atd_txdot_crashes(where: {crash_id: {_eq:$crash_id}}) {
+            cr3_file_metadata
+          }
+        }
+    """
+    response = requests.post(
+        url=os.getenv("HASURA_ENDPOINT"),
+        headers={
+            "Accept": "*/*",
+            "content-type": "application/json",
+            "x-hasura-admin-secret": os.getenv("HASURA_ADMIN_KEY")
+        },
+        json={
+            "query": query_get_record,
+            "variables": {
+                "crash_id": int(crash_id)
+            }
+        }
+    )
+    response.encoding = "utf-8"
+    try:
+        return response.json()["data"]["atd_txdot_crashes"][0]["cr3_file_metadata"]
+    except (IndexError, TypeError, KeyError):
```

Comment:
when would a `TypeError` be raised? may need to handle a JSON decoder error?

Reply:
Ah, good call. Thanks @johnclary